components/brave_sync/brave_profile_sync_service_impl.cc

```diff
@@ -198,25 +186,103 @@ void FillObjectsMap(const bookmarks::BookmarkNode* parent,
   }
 }
 
+void GetTopMostFolders(const NodesSet* nodes_with_duplicates,
+                       NodesSet* top_most_folders) {
+  for (const bookmarks::BookmarkNode* node : *nodes_with_duplicates) {
+    if (node->is_folder()) {
+      // Only folders are interesed here
+      bool set_top_node = false;
+      bool node_under_top = false;
+      NodesSet nodes_to_remove;
+      for (const bookmarks::BookmarkNode* top_folder_entry :
+           *top_most_folders) {
+        if (top_folder_entry->HasAncestor(node)) {
+          // node
+          //  L..
+          //    |
+          //   ...
+          //    L top_folder_entry
+          // top_folder_entry is not a top most anymore
+          nodes_to_remove.insert(top_folder_entry);
+          set_top_node = true;
+        } else if (node->HasAncestor(top_folder_entry)) {
+          // top_folder_entry
+          //  L..
+          //    |
+          //   ...
+          //    L node
+          node_under_top = true;
+          break;
+        }
+      }
+
+      for (const auto* node_to_remove : nodes_to_remove) {
+        top_most_folders->erase(node_to_remove);
+      }
+
+      if (set_top_node || !node_under_top) {
+        // We have duplicated folder which is above already detected top node
+        // Or we found a new top node - the node not yet seen as child of
+        // detected top node
+        top_most_folders->insert(node);
+      }
+    }
+  }
+}
+
+void FilterOutNestedNodes(NodesSet* nodes_with_duplicates,
+                          NodesSet* top_most_folders,
+                          NodesSet* filtered_nodes_with_duplicates) {
+  for (const bookmarks::BookmarkNode* node : *nodes_with_duplicates) {
+    bool include_into_delete = true;
+    for (const bookmarks::BookmarkNode* top_folder : *top_most_folders) {
+      if (top_folder != node && node->HasAncestor(top_folder)) {
+        // Don`t include the node if we are going to remove the parent folder
+        // Top folder must be included
+        include_into_delete = false;
+      }
+    }
+
+    if (include_into_delete) {
+      filtered_nodes_with_duplicates->insert(node);
+    }
+  }
+}
+
 void ClearDuplicatedNodes(ObjectIdToNodes* object_id_nodes,
                           bookmarks::BookmarkModel* model) {
+  size_t nodes_recreated = 0;
+  NodesSet nodes_with_duplicates;
   for (ObjectIdToNodes::iterator it_object_id = object_id_nodes->begin();
        it_object_id != object_id_nodes->end(); ++it_object_id) {
-    const SortedNodes& nodes = it_object_id->second;
+    const NodesSet& nodes = it_object_id->second;
     if (nodes.size() > 1) {
-      // Re-create all group of nodes which have the same object_id
-      for (SortedNodes::iterator it_nodes = nodes.begin();
-           it_nodes != nodes.end(); ++it_nodes) {
-        const bookmarks::BookmarkNode* node = *it_nodes;
```

Comment:
Let's simply the de-nesting process to avoid unnecessary iterations.
```cpp
NodesSet deleted_nodes;
if (deleted_nodes.find(node) != deleted_nodes.end())
  continue;
if (node->is_folder()) {
  // insert all descendents into the set
}
// Do the rest of Copy & Remove
```